examples/2_advanced/plot_s4_galaxies.py

```diff
@@ -0,0 +1,166 @@
+n_arcmin2 = 0.3
+
+# sigma_ellipticity as expected for a stage-IV survey
+sigma_e = 0.27
+
+# photometric redshift error
+sigma_z0 = 0.03
+
+# set up CAMB parameters for matter angular power spectrum
+pars = camb.set_params(H0=100*cosmo.h, omch2=cosmo.Om*cosmo.h**2)
+
+# %%
+# Simulation Setup
+# ----------------
+# Here we setup the overall photometric redshift distribution
+# and separate it into equal density tomographic bins
+# with photometric redshift errors.
+
+# setting up the random number generator:
+rng = np.random.default_rng(seed=42)
+
+# photometric redshift distribution following a Smail distribution
+z = np.linspace(0, 3.0, 1000)
+dndz = glass.observations.smail_nz(z, z_mode=0.9, alpha=2., beta=1.5)
+dndz *= n_arcmin2
+bz = 1.2
+
+# equal density bins:
+nbins = 10
+zedges = glass.observations.equal_dens_zbins(z, dndz, nbins=nbins)
+bin_nz = glass.observations.tomo_nz_gausserr(z, dndz, sigma_z0, zedges)
+
+# %%
+# Plotting the overall redshift distribution and the
+# distribution for each of the equal density tomographic bins
+plt.ion()
+plt.figure(figsize=(10, 5))
+plt.title("Stage IV Space Telescope - Photometric Distribution: equal density bins")
+SumNz = np.zeros_like(bin_nz[0])
+for nz in bin_nz:
+    plt.fill_between(z, nz, alpha=0.5)
+    SumNz = SumNz + nz
+plt.fill_between(z, dndz, alpha=0.2, label='dn/dz')
+plt.plot(z, SumNz, ls='--', label="Sum of the bins")
+plt.ylabel("dN/dz - gal/arcmin2")
+plt.xlabel("z")
+plt.legend()
+plt.tight_layout()
+plt.pause(1e-3)
+
+# %%
+# Make a visibility map with low NSIDE
+# also compute its fsky for the extected galaxy count
+stageIV_mask = glass.observations.vmap_galactic_ecliptic(nside)
+
+# checking the mask:
+hp.mollview(stageIV_mask, title='Stage IV Space Survey-like Mask', unit='Visibility')
+plt.pause(1e-3)
+
+# %%
+# generators for the clustering and lensing
+generators = [
+    glass.sim.zspace(0., 3.0001, dz=dz),
+    glass.camb.camb_matter_cl(pars, lmax),
+    glass.matter.lognormal_matter(nside, rng=rng),
+    glass.lensing.convergence(cosmo),
+    glass.lensing.shear(lmax),
+    glass.observations.vis_constant(stageIV_mask, nside=nside),
+    glass.galaxies.gal_dist_fullsky(z, bin_nz, bz=bz, rng=rng),
+    glass.galaxies.gal_ellip_gaussian(sigma_e, rng=rng),
+    glass.galaxies.gal_shear_interp(cosmo),
+]
+
+# %%
+# Simulation
+# ----------
+# Simulate the galaxies with shears.  In each iteration, get the quantities of interest
+# to build our mock catalogue.
+
+# keep count of total number of galaxies
+num = 0
+
+# we will store the catalogue as a dictionary:
+catalogue = {'RA': np.array([]), 'DEC': np.array([]), 'TRUE_Z': np.array([]),
+             'E1': np.array([]), 'E2': np.array([]), 'TOMO_ID': np.array([])}
+
+# iterate and store the quantities of interest for our mock catalogue:
+for shell in glass.sim.generate(generators):
+    print(f"Generating shell #: {shell['#']}")
+    num += shell['ngal']
```

Comment:
No need to print shell number of the loop; this actually makes it hard to see the total number of galaxies. Also no need to keep track of number of galaxies, the length of the catalogue does that.

```suggestion
for shell in glass.sim.generate(generators):
```